adapters/mediago/mediago.go

```diff
@@ -0,0 +1,222 @@
+	if err == nil {
+		adapterRequests = append(adapterRequests, adapterRequest)
+	} else {
+		errs = append(errs, err)
+	}
+	return adapterRequests, errs
+}
+
+func (a *adapter) makeRequest(request *openrtb2.BidRequest) (*adapters.RequestData, error) {
+	mediagoExt, err := getImpMediaGoExt(&request.Imp[0])
+
+	if err != nil {
+		return nil, &errortypes.BadInput{Message: "Invalid ExtImpMediaGo value"}
+	}
+
+	endPoint, err := a.getEndPoint(mediagoExt)
+	if err != nil {
+		return nil, err
+	}
+
+	preProcess(request)
+	reqBody, err := json.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("x-openrtb-version", "2.5")
+
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     endPoint,
+		Body:    reqBody,
+		Headers: headers,
+		ImpIDs:  openrtb_ext.GetImpIDs(request.Imp),
+	}, nil
+}
+
+// get ImpMediaGoExt From First Imp. Only check and get first Imp.Ext.Bidder to ExtImpMediago
+func getImpMediaGoExt(imp *openrtb2.Imp) (*openrtb_ext.ExtImpMediaGo, error) {
+	var extImpMediaGo openrtb_ext.ExtImpMediaGo
+	var extBidder adapters.ExtImpBidder
+	err := json.Unmarshal(imp.Ext, &extBidder)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(extBidder.Bidder, &extImpMediaGo)
+	if err != nil {
+		return nil, err
+	}
+	return &extImpMediaGo, nil
+}
+
+func getRegionInfo(region string) string {
+	switch region {
+	case "APAC":
+		return "jp"
+	case "EU":
+		return "eu"
+	case "US":
+		return "us"
+	default:
+		return "us"
+	}
+}
+
+func (a *adapter) getEndPoint(ext *openrtb_ext.ExtImpMediaGo) (string, error) {
+	endPointParams := macros.EndpointTemplateParams{
+		AccountID: url.PathEscape(ext.Token),
+		Host:      url.PathEscape(getRegionInfo(ext.Region)),
+	}
+	return macros.ResolveMacros(a.EndpointTemplate, endPointParams)
+}
+
+func preProcess(request *openrtb2.BidRequest) {
+	for i := range request.Imp {
+		if request.Imp[i].Banner != nil {
+			banner := *request.Imp[i].Banner
+			if (banner.W == nil || banner.H == nil || *banner.W == 0 || *banner.H == 0) && len(banner.Format) > 0 {
+				firstFormat := banner.Format[0]
+				banner.W = &firstFormat.W
+				banner.H = &firstFormat.H
+				request.Imp[i].Banner = &banner
+			}
+		}
+	}
+}
+
+func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d.", response.StatusCode),
+		}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d.", response.StatusCode),
+		}}
+	}
```

Comment:
could use of response util - https://github.com/prebid/prebid-server/blob/ec729e6e5d50a1306d76332fda52108b9d313c1d/adapters/response.go#L10-L28